src/client.rs

```diff
@@ -21,56 +26,124 @@ pub enum Client {
+                Err(Error::Protocol(_)) => {
+                    return res;
+                },
+                Err(e) => {
+                    warn!("call retry:{}/{} {:?}", errors.len() + 1 , $self.config.retry(), e);
+                    errors.push(e);
+                    if errors.len() as u8 == $self.config.retry() {
+                        return Err(Error::AllAttemptsErrored(errors));
+                    }
+
+                    // Only one thread will try to recreate the client getting the write lock,
+                    // other eventual threads will get Err and will block at the beginning of
+                    // previous loop when trying to read()
+                    if let Ok(mut write_client) = $self.client_type.try_write() {
+                        loop {
+                            std::thread::sleep(std::time::Duration::from_secs(errors.len() as u64));
+                            match ClientType::from_config(&$self.url, &$self.config) {
+                                Ok(new_client) => {
+                                    info!("Succesfully created new client");
+                                    *write_client = new_client;
+                                    break;
+                                },
+                                Err(e) => {
+                                    warn!("client retry:{}/{} {:?}", errors.len() + 1, $self.config.retry(), e);
+                                    errors.push(e);
+                                    if errors.len() as u8 == $self.config.retry() {
+                                        return Err(Error::AllAttemptsErrored(errors));
+                                    }
+                                }
+                            }
+                        }
+                    }
+                },
+            }
+        }}
+    }
+}
+
+impl ClientType {
+    ///  Constructor that supports multiple backends and allows configuration through
+    /// the [Config]
+    pub fn from_config(url: &str, config: &Config) -> Result<Self, Error> {
+        if url.starts_with("ssl://") {
+            if config.socks5().is_some() {
+                return Err(Error::SSLOverSocks5);
+            }
+
+            let url = url.replacen("ssl://", "", 1);
+            let client =
+                RawClient::new_ssl(url.as_str(), config.validate_domain(), config.timeout())?;
+            Ok(ClientType::SSL(client))
+        } else {
+            let url = url.replacen("tcp://", "", 1);
+
+            Ok(match config.socks5().as_ref() {
+                None => ClientType::TCP(RawClient::new(url.as_str(), config.timeout())?),
+                Some(socks5) => ClientType::Socks5(RawClient::new_proxy(url.as_str(), socks5)?),
+            })
         }
     }
 }
 
 impl Client {
-    /// Generic constructor that supports multiple backends and, optionally, a socks5 proxy.
+    /// Default constructor supporting multiple backend by providing a prefix
```

Comment:
Typo: "... supporting multiple backends ..."

Reply:
fixed in f0e863f1cda111ec746ed2d95923db234e85d8fd